runtime/compiler/aarch64/runtime/ARM64RelocationTarget.cpp

```diff
@@ -27,67 +27,93 @@ void arm64CodeSync(unsigned char *codeStart, unsigned int codeSize);
    {
-   TR_UNIMPLEMENTED();
-   return 0;
+   // reloLocation points at the cpAddress in snippet, need to find cpIndex location
+   reloLocation += sizeof(uintptr_t);
+   return (uintptr_t )loadPointer(reloLocation);
    }
 
 void
 TR_ARM64RelocationTarget::performThunkRelocation(uint8_t *thunkBase, uintptr_t vmHelper)
    {
-   TR_UNIMPLEMENTED();
+   uintptr_t sendTarget = vmHelper;
+   int32_t *thunkRelocationData = (int32_t *)(thunkBase - sizeof(int32_t));
+
+   storeAddressSequence((uint8_t *)vmHelper, thunkBase + *thunkRelocationData, 1);
+
+   arm64CodeSync(thunkBase, *((int32_t*)(thunkBase - 2*sizeof(int32_t))));
    }
 
 uint8_t *
 TR_ARM64RelocationTarget::loadAddressSequence(uint8_t *reloLocation)
    {
-   TR_UNIMPLEMENTED();
-   return NULL;
+   TR_ASSERT(0, "AOT New Relo Runtime: don't expect loadAddressSequence to be called!\n");
+   uintptr_t value = 0;
+   return (uint8_t *)value;
    }
 
 void
 TR_ARM64RelocationTarget::storeAddressSequence(uint8_t *address, uint8_t *reloLocation, uint32_t seqNumber)
    {
-   TR_UNIMPLEMENTED();
+   uintptr_t value = (uintptr_t)address;
+   uint32_t *movInstruction = (uint32_t *) reloLocation;
+   // immediate offset is at bit 20:5
+   *movInstruction = (*movInstruction & (~((uint32_t)0xFFFF << 5))) | ((value & 0x0000FFFF) << 5);
+   *(movInstruction + 1) = (*(movInstruction + 1) & (~((uint32_t)0xFFFF << 5))) | (((value >> 16) & 0x0000FFFF) << 5);
+   *(movInstruction + 2) = (*(movInstruction + 2) & (~((uint32_t)0xFFFF << 5))) | (((value >> 32) & 0x0000FFFF) << 5);
+   *(movInstruction + 3) = (*(movInstruction + 3) & (~((uint32_t)0xFFFF << 5))) | (((value >> 48) & 0x0000FFFF) << 5);
    }
 
 void
 TR_ARM64RelocationTarget::storeRelativeTarget(uintptr_t callTarget, uint8_t *reloLocation)
    {
-   TR_UNIMPLEMENTED();
+   // reloLocation points at the start of the call offset, so just store the uint8_t * at reloLocation
+   int32_t instruction = *((int32_t *)reloLocation);
+
+   instruction &= (int32_t)0xFC000000;
+   // labels will always be 4 aligned, CPU shifts # left by 2
+   // during execution (26 bits -> 28 bit branch reach)
+   instruction |= (callTarget >> 2) & 0x03FFFFFF;
+
+   storeSigned32b(instruction, (uint8_t *)reloLocation);
    }
 
 bool
-TR_ARM64RelocationTarget::useTrampoline(uint8_t * helperAddress, uint8_t *baseLocation)
+TR_ARM64RelocationTarget::useTrampoline(uint8_t *helperAddress, uint8_t *baseLocation)
    {
-   TR_UNIMPLEMENTED();
-   return false;
+   return
+      !TR::Compiler->target.cpu.isTargetWithinUnconditionalBranchImmediateRange((intptrj_t)helperAddress, (intptrj_t)baseLocation) ||
+      TR::Options::getCmdLineOptions()->getOption(TR_StressTrampolines);
    }
 
 uint8_t *
 TR_ARM64RelocationTarget::arrayCopyHelperAddress(J9JavaVM *javaVM)
    {
-   TR_UNIMPLEMENTED();
-   return NULL;
+   uintptr_t *funcdescrptr = (UDATA *)javaVM->memoryManagerFunctions->referenceArrayCopy;
+   uintptr_t value = (UDATA)funcdescrptr;
+
+   return (uint8_t *)value;
```

Comment:
Might be more concise if it was just:
```
   return reinterpret_cast<uint8_t *>(javaVM->memoryManagerFunctions->referenceArrayCopy);
```

Reply:
I agree. I updated the function as proposed.